processing/src/main/java/io/druid/query/QueryMetricsFactory.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query;
+
+import io.druid.query.groupby.GroupByQueryMetrics;
+import io.druid.query.timeseries.TimeseriesQueryMetrics;
+import io.druid.query.topn.TopNQueryMetrics;
+
+/**
+ * This factory is used for DI of custom {@link QueryMetrics} implementations.
+ */
+public interface QueryMetricsFactory
+{
+  /**
+   * Creates a generic {@link QueryMetrics} for query, which doesn't require a custom QueryMetrics subclass (i. e. not
+   * a topN, groupBy or timeseries query). This method must call {@link QueryMetrics#query(Query)} with the given query
+   * on the created QueryMetrics object before returning.
+   */
+  QueryMetrics<Query<?>> makeMetrics(Query<?> query);
```

Comment:
I don't like the idea of adding XxxQueryMetricsFactory for each existing query type up front, while they don't yet need a custom `QueryMetrics` subtype. Added dedicated factories for topN, groupBy and timeseries.

For total compatibility in all directions a complex plan of moving from generic `QueryMetrics` to a query-specific subtype should be followed. I put this plan in `QueryMetrics` javadocs, "Making subinterfaces of QueryMetrics" section. Please review this plan.